Kind and sizes

ZedThree opened the pull request:
Built on top of #87 to avoid massive conflicts.

Closes #86 

- Separate kind suffixes on number literals so they appear in the syntax tree
- Rename the `size` node on types as `kind` to distinguish it from array sizes
- Parse sizes on variable declarations as `size` instead of `call_expression`
- Add `variable` node in `variable_declaration` to group the variable name and array size

ZedThree commented:
@stadelmanma I finally revisited this to rebase it onto `master`.

I'm not sure about the `variable` node -- it does make it nicer when grouping name + size for arrays, but it's not needed for scalar variables so it's a bit more noise in the tree. Thoughts?

I do think capturing the kinds on literals is useful though, and we should add that

ZedThree commented:
I also have something for parsing string literals with kinds, except that it eats the `_` as part of the kind identifier, which is a little annoying

stadelmanma commented:
> @stadelmanma I finally revisited this to rebase it onto `master`.
> 
> I'm not sure about the `variable` node -- it does make it nicer when grouping name + size for arrays, but it's not needed for scalar variables so it's a bit more noise in the tree. Thoughts?
> 
> I do think capturing the kinds on literals is useful though, and we should add that

I think the benefit of having `(variable (identifier) (size (number_literal)))` is worth the added noise you get when seeing `(variable (identifier))` in the parse tree. It also adds a level of explicitness to the output vs just `identifier` which can appear in a lot of different contexts.

> I also have something for parsing string literals with kinds, except that it eats the _ as part of the kind identifier, which is a > little annoying

Is this already in the PR as well?

ZedThree commented:
> I think the benefit of having (variable (identifier) (size (number_literal))) is worth the added noise you get when seeing (variable (identifier)) in the parse tree. It also adds a level of explicitness to the output vs just identifier which can appear in a lot of different contexts.

I've just had a look at what the C grammar does, because they have something similar with e.g. `int foo[3];`, and they have this:

```js
    declaration: $ => seq(
      $._declaration_specifiers,
      commaSep1(field('declarator', choice(
        seq(
          optional($.ms_call_modifier),
          $._declaration_declarator,
          optional($.gnu_asm_expression),
        ),
        $.init_declarator,
      ))),
      ';',
    ),

// ...

    _declaration_declarator: $ => choice(
      $.attributed_declarator,
      $.pointer_declarator,
      alias($._function_declaration_declarator, $.function_declarator),
      $.array_declarator,
      $.parenthesized_declarator,
      $.identifier,
    ),

// ...

    array_declarator: $ => prec(1, seq(
      field('declarator', $._declarator),
      '[',
      repeat(choice($.type_qualifier, 'static')),
      field('size', optional(choice($.expression, '*'))),
      ']',
    )),
```

With an example tree:

```
int foo, zing;
int bar[3];
---
(translation_unit
  (declaration
    type: (primitive_type)
    declarator: (identifier)
    declarator: (identifier))
  (declaration
    type: (primitive_type)
    declarator: (array_declarator
      declarator: (identifier)
      size: (number_literal))))
```

I think this is better than my attempt here. I'll try and rework it to use this sort of structure.

> Is this already in the PR as well? I don't think I'm following the _ getting consumed.

No, sorry, not yet. I'm not really happy with how it works eating the `_`. To be clear, the issue is something like this:

```fortran
foo = unicode_"Some string"
```

where `unicode` is the string literal kind specifier, and is some identifier. Whitespace is not allowed either side of the underscore. There's two issues why this isn't straightforward:

1. `unicode_` is of course a valid identifier, so this is difficult to parse without capturing the trailing underscore as part of it;
2. the string literal rule is an external scanner, and `token.immediate($.string_literal)` isn't allowed (the immediate rule must be a plain regexp or string for reasons).

I think 2. isn't such an issue -- parsing invalid Fortran is better than not parsing valid Fortran -- but I'm struggling to get something working for 1 at all

ZedThree commented:
Ok, this took way, way too much effort for the actual payoff, but I now have working string literal kind parameters. We're now parsing 99% of files without error (97% if preprocessor files are included).

Here's a flavour:

```fortran
program foo
integer :: f_1 = 1_int64
character(len=*), parameter :: a = ck_"hi""bye"

end program foo
```

parses to:

```
(translation_unit [0, 0] - [5, 0]
  (program [0, 0] - [5, 0]
    (program_statement [0, 0] - [0, 11]
      (name [0, 8] - [0, 11]))
    (variable_declaration [1, 0] - [1, 24]
      type: (intrinsic_type [1, 0] - [1, 7])
      declarator: (init_declarator [1, 11] - [1, 24]
        left: (identifier [1, 11] - [1, 14])
        right: (number_literal [1, 17] - [1, 24]
          kind: (identifier [1, 19] - [1, 24]))))
    (variable_declaration [2, 0] - [2, 47]
      type: (intrinsic_type [2, 0] - [2, 9])
      type: (kind [2, 9] - [2, 16]
        (argument_list [2, 9] - [2, 16]
          (keyword_argument [2, 10] - [2, 15]
            name: (identifier [2, 10] - [2, 13])
            value: (assumed_size [2, 14] - [2, 15]))))
      attribute: (type_qualifier [2, 18] - [2, 27])
      declarator: (init_declarator [2, 31] - [2, 47]
        left: (identifier [2, 31] - [2, 32])
        right: (string_literal [2, 35] - [2, 47]
          kind: (identifier [2, 35] - [2, 37]))))
    (end_program_statement [4, 0] - [5, 0]
      (name [4, 12] - [4, 15]))))
```

And a query:
```
(number_literal) @literal
(number_literal kind: (identifier) @kind)
(string_literal kind: (identifier) @kind)
(string_literal ) @literal
```
returns:
```
string_literal.f90
  pattern: 0
    capture: 0 - literal, start: (1, 17), end: (1, 24), text: `1_int64`
  pattern: 1
    capture: 1 - kind, start: (1, 19), end: (1, 24), text: `int64`
  pattern: 3
    capture: 0 - literal, start: (2, 35), end: (2, 47), text: `ck_"hi""bye"`
  pattern: 2
    capture: 1 - kind, start: (2, 35), end: (2, 37), text: `ck`
```

The only annoying thing is that the literals include their kind parameters, but I think adding another node would end up with way too much noise. At least now it should be both easy and consistent to strip the kind from the literal.

As a side note, the C and C++ grammars _do_ have another node for the string literal contents. Maybe something to consider

ZedThree commented:
Ok, trying not to fiddle with this any more! I just removed the `(argument_list)` node from appearing in `(kind)`. Basically, parse the following:

    real(real64) :: foo

as

    (variable_declaration
      (intrinsic_type)
      (kind
        (identifier))
      (identifier))

instead of

    (variable_declaration
      (intrinsic_type)
      (kind
        (argument_list
          (identifier)))
      (identifier))

Then I also added parameterised types, because they're actually related to this, and then fixed one minor issue with derived types.

In terms of language features, I think this really only leaves coarrays left.

ZedThree commented:
Having used this in practice a little, I'm still not completely happy:

```fortran
real(real64) :: foo = 1.0_real64
```

gets parsed as

```text
    (variable_declaration
      type: (intrinsic_type)
      type: (kind
        (identifier))
      declarator: (init_declarator
        left: (identifier)
        right: (number_literal
          kind: (identifier))))
```

with the type's kind appearing as a sibling node to the actual type, while the literal's kind is a child (and a named field).

Although that won't work so well for `characters`:

```fortran
character(len=5, kind=c_char) :: foo = c_char_"hello"
```

Currently with this PR:

```text
    (variable_declaration
      type: (intrinsic_type)
      type: (kind
        (keyword_argument
          name: (identifier)
          value: (number_literal))
        (keyword_argument
          name: (identifier)
          value: (identifier)))
      declarator: (init_declarator
        left: (identifier)
        right: (string_literal
          kind: (identifier))))
```

So either as fields:


```text
    (variable_declaration
      type: (intrinsic_type
        kind: (keyword_argument
          name: (identifier)
          value: (number_literal))
        kind: (keyword_argument
          name: (identifier)
          value: (identifier)))
      declarator: (init_declarator
        left: (identifier)
        right: (string_literal
          kind: (identifier))))
```

or as a child node of the type:

```text
    (variable_declaration
      type: (intrinsic_type
        (kind
          (keyword_argument
            name: (identifier)
            value: (number_literal))
          (keyword_argument
            name: (identifier)
            value: (identifier))))
      declarator: (init_declarator
        left: (identifier)
        right: (string_literal
          kind: (identifier))))
```

@stadelmanma thoughts on either of those options? or something else?

ZedThree commented:
After playing around with trying to use various versions, I've settled on this:

```
    (variable_declaration
      type: (intrinsic_type
        kind: (kind
          (keyword_argument
            name: (identifier)
            value: (number_literal))
          (keyword_argument
            name: (identifier)
            value: (identifier))))
      declarator: (init_declarator
        left: (identifier)
        right: (string_literal
          kind: (identifier))))
```

where there is both a `kind` node _and_ a `kind` field -- this makes it really easy to pull out the kind, while also wrapping up children nodes.

Derived types are now also treated symmetrically too:

```fortran
type(character(len=5, kind=c_char)) :: foo = c_char_"hello"
```

```
    (variable_declaration
      type: (derived_type
        name: (intrinsic_type)
        kind: (kind
          (keyword_argument
            name: (identifier)
            value: (number_literal))
          (keyword_argument
            name: (identifier)
            value: (identifier))))
      declarator: (init_declarator
        left: (identifier)
        right: (string_literal
          kind: (identifier))))
```

This gives us a `type` field that has a child `kind` field, which means this can be treated similarly to intrinsic types. It is maybe a little weird that `(kind)` is a sibling of `(intrinsic_type)` here instead of a child, but I think that's a reasonable compromise.

@stadelmanma Ok to merge and make a new release?

stadelmanma commented:
I think that looks great, I had thought I responded to this thread about the initially odd node structure but you implemented exactly what I was thinking of. I'll merge this now, might need to wait to cut a release until I'm home.

stadelmanma commented:
@ZedThree new release is cut, not sure if you use the packages but pypi and crates are updated. I still need to fiddle with NPM one day.

ZedThree commented:
Amazing, thanks! I've already managed to start using the new release (in a linter my colleague is developing, [fortitude](https://github.com/PlasmaFAIR/fortitude))

Now I need to update the fixed-form grammar to use the new release (and tree-sitter 0.23)...
